spec/abstract/IpcCancel_A.thy

```diff
@@ -300,37 +300,20 @@ where
     BlockedOnNotification r \<Rightarrow> Some r
   | _ \<Rightarrow> None"
 
-fun
-  tcb_ep_find_index :: "obj_ref \<Rightarrow> obj_ref list \<Rightarrow> nat \<Rightarrow> (nat, 'z::state_ext) s_monad"
-where
-  "tcb_ep_find_index tptr qs curindex = do
-     prio \<leftarrow> thread_get tcb_priority tptr;
-     curprio \<leftarrow> thread_get tcb_priority (qs ! curindex);
-     if prio > curprio then
-       if curindex = 0 then return 0
-       else tcb_ep_find_index tptr qs (curindex - 1)
-     else return (curindex + 1)
-   od"
+definition tcb_ep_dequeue :: "obj_ref \<Rightarrow> obj_ref list \<Rightarrow> (obj_ref list, 'z::state_ext) s_monad" where
+  "tcb_ep_dequeue tptr qs \<equiv> return $ filter ((\<noteq>) tptr) qs"
 
-declare tcb_ep_find_index.simps[simp del]
-
-definition
-  tcb_ep_dequeue :: "obj_ref \<Rightarrow> obj_ref list \<Rightarrow> (obj_ref list, 'z::state_ext) s_monad"
-where
-  "tcb_ep_dequeue tptr qs = do
-     index \<leftarrow> return $ the $ findIndex (\<lambda>x. x = tptr) qs;
-     return $ take index qs @ drop (index + 1) qs
+definition tcb_ep_append :: "obj_ref \<Rightarrow> obj_ref list \<Rightarrow> (obj_ref list, 'z::state_ext) s_monad" where
+  "tcb_ep_append tptr qs \<equiv> do
+     prio \<leftarrow> thread_get tcb_priority tptr;
+     prios \<leftarrow> mapM (thread_get tcb_priority) qs;
+     zprios \<leftarrow> return $ zip qs prios;
+     zprios' \<leftarrow> return $ filter (\<lambda>(_, p). p \<ge> prio) zprios
+                         @ [(tptr, prio)]
+                         @ filter (\<lambda>(_, p). p < prio) zprios;
+     return (map fst zprios')
    od"
```

Comment:
comprehension nitpick: if it's placing the tptr into a list such that it's sorted, does it still make sense to call it "append" without comment?

Reply:
Good point, it'd be more like an `insert`. I suspect the name hasn't changed in C either, but we could still do better in the specs.